event/metrics.go

```diff
@@ -0,0 +1,91 @@
+package event
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// MustRegisterMetrics will register all event related metrics on the given registry.
+// If metrics with the same name already exist no the register this function will panic.
+func MustRegisterMetrics(registry *prometheus.Registry) {
+	registry.MustRegister(publishDuration, publishCounter, processCounter, processDuration)
+}
+
+// SampledMessageHandler will instrument the given MessageHandler returning a new one
+// that samples metrics. These will be `event_process_*` metrics using as `name` the
+// given eventName.
+func SampledMessageHandler(handler MessageHandler, eventName string) MessageHandler {
+	return func(msg Message) error {
+		start := time.Now()
+		err := handler(msg)
+		elapsed := time.Since(start)
+		sampleProcess(eventName, elapsed, err)
+		return err
+	}
+}
+
+func samplePublish(name string, elapsed time.Duration, err error) {
+	status := "ok"
+	if err != nil {
+		status = "error"
+	}
+	labels := prometheus.Labels{
+		"status": status,
+		"name":   name,
+	}
+	publishDuration.With(labels).Observe(elapsed.Seconds())
+	publishCounter.With(labels).Inc()
+}
+
+func sampleProcess(name string, elapsed time.Duration, err error) {
+	status := "ok"
+	if err != nil {
+		status = "error"
+	}
+	labels := prometheus.Labels{
+		"status": status,
+		"name":   name,
+	}
+	processDuration.With(labels).Observe(elapsed.Seconds())
+	processCounter.With(labels).Inc()
+}
+
+var (
```

Comment:
Doubt: Is it normal to define var in the end of the file?

Reply:
Not normal, there is no default/standard way to do this in Go. I like starting the file with everything that is public and leaving everything that is private after the public things, but that is not something that everyone does..it is something that I like. I'm a very top/down person and when I read code I like to see public interfaces/types/functions first and then to start digging in to internals. But again, not a normal thing...just my thing =P.

Reply:
I also like this approach actually... Most important things on top and going down with more specific/less important... Also I like to define methods related to a struct close to where the struct is defined